tests/abi/abi.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * This file is distributed under the MIT License. See LICENSE.md for details.
+ */
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "for_each.h"
+#include "functions.h"
+
+#ifdef __x86_64__
+#include "platform_specific/x86_64.h"
+#elif __x86__
+#include "platform_specific/x86.h"
+#else
+_Static_assert(0, "unsupported platform architecture");
+#endif
+
+/* LFSR generator */
+
+static uint64_t LFSR = 0xdc984ef2647af904ull;
+uint64_t getNextLFSR(void) {
+  uint32_t Iteration;
+  for (Iteration = 0; Iteration < 64; ++Iteration)
+    LFSR = (LFSR >> 1) | ((((LFSR >> 1) ^ (LFSR >> 2)) & 1) << 63);
+  return LFSR;
+}
+
+/* Helpers */
+
+#define GET_1ST_HELPER(THIS_ONE, ...) THIS_ONE
+#define GET_1ST(THIS_ONE, ...) GET_1ST_HELPER(THIS_ONE, __VA_ARGS__)
+#define FIRST_REGISTER GET_1ST(REGISTER_LIST)
+
+#define STACK_VALUE_COUNT 32
+#define REGISTER_COUNT COUNT_ARGUMENTS(REGISTER_LIST)
+#define GENERATED_COUNT REGISTER_COUNT + STACK_VALUE_COUNT
+
+void generate(REGISTER_C_TYPE *Generated) {
```

Comment:
* [ ] Can you add a `count` argument instead of hardcoding `GENERATED_COUNT`? Alternatively, make a `struct CPUState` containing an array with the right number of elements.
* [ ] Maybe `generate_random` is a more appropriate name.

Reply:
1. The `GENERATED_COUNT` is also extensively used at the point which the numbers are generated for. That's why I went with the macro, and I can't afford to spend another register to hold the count at the point where it's needed later down the line, so it has to be defined at compile time.
2. It was a struct initially, I don't remember exactly what ticked me off to go to the inferior "pointer-to-array" approach, but something about the assembly it was generating was messing up the stuff later down the line.
3. I agree about the name.

Reply:
1 & 2. Irrelevant since it became a global variable.
3. I chose to rename the whole thing into `random_state`, `regenerate_state` and `print_current_state`, what do you think?